ledger/src/ui/src/onboard.c

```diff
@@ -0,0 +1,160 @@
+    // hasn't been done just in case something fails
+    nvm_write((void *)PIC(N_onboarded_ui),
+              (void *)&onboarded_flag,
+              sizeof(onboarded_flag));
+
+#ifndef DEBUG_BUILD
+    if (!is_pin_valid(pin_ctx)) {
+        THROW(ERR_INVALID_PIN);
+    }
+#endif
+
+    // Wipe device
+    os_global_pin_invalidate();
+    os_perso_wipe();
+    // Generate 32 bytes of random with onboard rng
+    cx_rng((unsigned char *)onboard_ctx->client_seed, HASHSIZE);
+    // XOR with host-generated 32 bytes random
+    for (int i = 0; i < HASHSIZE; i++) {
+        onboard_ctx->client_seed[i] ^= onboard_ctx->words_buffer[i];
+    }
+    // The seed is now in client_seed, generate the mnemonic
+    os_memset(onboard_ctx->words_buffer, 0, sizeof(onboard_ctx->words_buffer));
+    onboard_ctx->words_buffer_length =
+        bolos_ux_mnemonic_from_data((unsigned char *)onboard_ctx->client_seed,
+                                    SEEDSIZE,
+                                    (unsigned char *)onboard_ctx->words_buffer,
+                                    sizeof(onboard_ctx->words_buffer));
+    // Clear the seed
+    explicit_bzero(onboard_ctx->client_seed, sizeof(onboard_ctx->client_seed));
+
+    // Set seed from mnemonic
+    os_perso_derive_and_set_seed(0,
+                                 NULL,
+                                 0,
+                                 NULL,
+                                 0,
+                                 onboard_ctx->words_buffer,
+                                 strlen(onboard_ctx->words_buffer));
+
+    // Clear the mnemonic
+    explicit_bzero(onboard_ctx->words_buffer,
+                   sizeof(onboard_ctx->words_buffer));
+
+    // Set PIN
+    os_perso_set_pin(0, GET_PIN(pin_ctx), GET_PIN_LENGTH(pin_ctx));
+    // Finalize onboarding
+    os_perso_finalize();
+    os_global_pin_invalidate();
+    unsigned char output_index = CMDPOS;
+    SET_APDU_AT(output_index++, 2);
+    SET_APDU_AT(output_index++,
+                os_global_pin_check(GET_PIN(pin_ctx), GET_PIN_LENGTH(pin_ctx)));
+
+    // Turn the onboarding flag on to mark onboarding
+    // has been done using the UI
+    onboarded_flag = 1;
+    nvm_write((void *)PIC(N_onboarded_ui),
+              (void *)&onboarded_flag,
+              sizeof(onboarded_flag));
+
+    return output_index;
+}
+
+/*
+ * Implement the RSK SEED command.
+ *
+ * Receives one byte at a time and fills words_buffer with the client-generated
+ * seed.
+ *
+ * @arg[in]  rx          number of received bytes from the Host
+ * @arg[out] onboard_ctx onboard context
+ * @ret                  number of transmited bytes to the host
+ */
+unsigned int set_client_seed(volatile unsigned int rx, onboard_t *onboard_ctx) {
+    // Should receive 1 byte per call
+    if (APDU_DATA_SIZE(rx) != 1) {
+        THROW(PROT_INVALID);
+    }
+
+    unsigned char index = APDU_OP();
+    if ((index >= 0) && ((size_t)index <= sizeof(onboard_ctx->words_buffer))) {
```

Comment:
I think this should fill `client_seed` and the `cx_rng` should fill another buffer called `seed`, that we then XOR into `seed`. That is then used to generate `words_buffer`. So then maybe `words_buffer` and `client_seed` can go into a `union`. Makes sense? Also we might find use in rewording "client" to "host". Up to you :)